feat(core): add vector tile loading in TMS

zarov opened the pull request:
Add vector tile loading in TMS protocol, with only Mapbox Vector Tile supported for now. It adds a new Parser (VectorTileParser), and other utility functions for vector tiles are in VectorTileHelper. 
Thanks to @peppsac for helping on this.

This is still in progress, as I have two things I want to clarify:
- should we keep the cache in `VectorTileHelper` ? If we drop it, it can drastically reduce the file.
- some things have been added for the styling in `Feature2Texture`, maybe it can be moved into another PR / commit ?

Also I need to add a correct `package-lock.json` file.

zarov commented:
I updated the PR, with two working examples. It seems like a big PR, but you can review it as two differents commits, the first one for the core and the other for the examples.

peppsac commented on examples/globe_vector_tile.js:
```diff
@@ -0,0 +1,132 @@
+        updateStrategy: {
+            type: itowns.STRATEGY_DICHOTOMY,
+        },
+        style: function (properties, feature) {
```
I'd move the mapbox-style converter to itowns core.

zarov replied:
I was thinking of maybe doing another PR introducing Stylizers ?

peppsac replied:
IMHO it's parsing / translating mapbox-style format, so you could already move this in `VectorTileHelper.js`

peppsac commented on examples/globe_vector_tile.js:
```diff
@@ -0,0 +1,132 @@
+
+            return styles;
+        },
+        filter: function (properties, geometry) {
```
Idem: the implementation of mapbox style filtering should go in core.

zarov replied:
The problem I encounter here is with the `mapboxLayers` global var: I can't figure how to pass properly this to internal methods in iTowns. Do you have a suggestion ?

peppsac replied:
- add a `style: https://raw.githubusercontent.com/Oslandia/postile-openmaptiles/master/style.json` to the layer
- fetch this json in the `preprocess` method of the provider. Then parse it and store the mapboxLayers in the layer?
WDYT?

zarov replied:
That could be one option yes, I'm trying it out

peppsac commented on examples/globe_vector_tile.js:
```diff
@@ -0,0 +1,132 @@
+
+
+exports.view = view;
+exports.initialPosition = positionOnGlobe;
```
Is this needed?

zarov replied:
Nope, bad copy paste

peppsac commented on examples/planar_vector_tile.js:
```diff
@@ -0,0 +1,180 @@
+viewerDiv.addEventListener('DOMMouseScroll', onMouseWheel);
+viewerDiv.addEventListener('mousewheel', onMouseWheel);
+
+viewerDiv.addEventListener('mousedown', function mouseDown(event) {
```
I suppose you'll want to use `PlanarControls` instead, like you did in #734 ?

zarov replied:
Yep, after merging 734 I'll remove this here.

peppsac commented on examples/planar_vector_tile.html:
```diff
@@ -0,0 +1,54 @@
+            #viewerDiv {
+                margin: auto auto;
+                width: 100%;
+                height: 100%;
```
We could add a `example.css` file instead of copy-pasting everytime :)

zarov replied:
Yes !

zarov replied:
Did #736 about that

mbredif commented on src/Parser/VectorTileParser.js:
```diff
@@ -0,0 +1,85 @@
+            case 'application/x-protobuf;type=mapbox-vector':
+            default:
+                return Promise.resolve(readPBF(file, options));
+        }
```
Should not you `Promise.reject` the default case ?

zarov replied:
I was on the assumption that if someone uses this parser, the file is always a vector tile, whatever the format, with the default case being being this one. So no rejection, as we trust the user to know what he does.

mbredif replied:
If you trust the user, then there is no use to switch on the format (maybe you have an extension in mind, but for now the switch is useless, just return the resolve directly).
If you want to advertise the mimetype, I propose in #723 to make it a field of the parsers.

zarov replied:
Yes, that's why I'm eager to see the parsers coming here :grinning:

mbredif commented on src/Parser/VectorTileParser.js:
```diff
@@ -0,0 +1,85 @@
+    });
+
+    return GeoJsonParser.parse(geojson, {
+        json: true,
```
the `json` option is not used anymore, please remove it ([it is now automatic](https://github.com/iTowns/itowns/blob/master/src/Parser/GeoJsonParser.js#L273-L275))

zarov replied:
Thanks !

mbredif commented on src/Provider/TMSProvider.js:
```diff
@@ -23,32 +24,56 @@ function preprocessDataLayer(layer) {
     for (const coordTMS of tile.getCoordsForLayer(layer)) {
         const coordTMSParent = (command.targetLevel < coordTMS.zoom) ?
             OGCWebServiceHelper.WMTS_WGS84Parent(coordTMS, command.targetLevel) :
             undefined;
```
Is it intentionnal that `coordTMSParent` is not used for vector tiles ? if yes, please move it to the OGC case.
My guess is that `const urld = URLBuilder.xyz(coordTMSParent || coordTMS, layer);` could be instead moved out here of the format switch (and of `getVectorTile`).

zarov replied:
Yes it is intentional. What do you mean by moving it to the OGC case ?

mbredif replied:
`URLBuilder.xyz` is called at two places with the same arguments : 
- L31 `const url = URLBuilder.xyz(coords, layer); `  (coords being `coordTMSParent || coordTMS`)
- L55 `const urld = URLBuilder.xyz(coordTMSParent || coordTMS, layer); `
You could move it right after L51 and pass the url to `VectorTileHelper.getVectorTileTextureByUrl` directly (removing the need for a `getVectorTile` function)

zarov replied:
Yeah I forgot to remove the `coordTMSParent ||` for the vector tile when I was testing before, if it is added tiles are not showing correctly.

mbredif replied:
Does it make sense to support `command.targetLevel` and `coordTMSParent` for vector tiles as well ? if not, why ? if yes, then I think this feature could be added to Vector tiles as well, that would help to have an equal treatment for OGCWebService, VectorTiles and any future formats... 

It is not a blocker to me, but if you merge it like this please open an issue about this for future reference.

mbredif commented on src/Provider/TMSProvider.js:
```diff
@@ -23,32 +24,56 @@ function preprocessDataLayer(layer) {
+                    texture,
+                };
+                return result;
+            });
```
could not this wrapping be done in `VectorTileHelper.getVectorTileTextureByUrl` and the `URLBuilder.xyz` be done outside of the format switch (see below), so as to remove this function entirely ?

zarov replied:
This wrapping could, but unfortunately the url no.

zarov replied:
A quick update: I need to keep this method if I want to support different type of layer (like `geometry`), as it is the only way for now ;)

mbredif reviewed (commented):
Looks good !
I do not really like the coupling of the `TMSProvider` introduced by the imports of parsers/helpers (`OGCWebServiceHelper` and `VectorTileHelper`) and of this local registry `supportedFormats`, but this is the current practice in itowns. IMHO, a parser/helper object should be instead provided in the layer options.

zarov commented:
Thanks !

> I do not really like the coupling of the TMSProvider introduced by the imports of parsers/helpers (OGCWebServiceHelper and VectorTileHelper) and of this local registry supportedFormats, but this is the current practice in itowns. IMHO, a parser/helper object should be instead provided in the layer options.

I agree, this is not ideal, but we can either merge this PR now, or wait for all the modifications we wants to be integrated in iTowns: it doesn't matter to me, I'm ok with waiting.

mbredif commented on src/Provider/TMSProvider.js:
```diff
@@ -23,32 +24,50 @@ function preprocessDataLayer(layer) {
+                }
+                return result;
+            }));
+        }
```
now that you use `coordTMSParent` if `func` is defined, dont you need to introduce this postprocessing (most notably `pitch`) in all cases ?

zarov commented:
@iTowns/reviewers I rebased this PR, this should be good to validate unless I missed something

peppsac commented on examples/globe_vector_tile.html:
```diff
@@ -0,0 +1,157 @@
+            var count = 0;
+            // Add a vector tile layer
+            itowns.Fetcher.json('https://raw.githubusercontent.com/Oslandia/postile-openmaptiles/master/style.json').then(function (style) {
+                // add one layer per layer in style.json
```
This comment is outdated and should be removed.

peppsac commented on src/Provider/TMSProvider.js:
```diff
@@ -23,32 +24,50 @@ function preprocessDataLayer(layer) {
     const tile = command.requester;
 
     const promises = [];
-    for (const coordTMS of tile.getCoordsForLayer(layer)) {
```
I'd prefer to keep this logic, and adapt the vector-tiles processing to fit in.

(it'll probably needs to use `Cache` to be able to re-use parent tile)

zarov replied:
I'll try to see what I can do.

peppsac commented on src/Provider/VectorTileHelper.js:
```diff
@@ -0,0 +1,77 @@
+export default {
+    /**
+     * Get a vector tile file, parse it and return a [Feature]{@link module:GeoJsonParser.Feature}
+     * or an array of Features. See [VectorTileParser]{@link module:VectorTileParser.parse}
```
I think that this comment is wrong since GeoJsonParser always returns a FeatureCollection

zarov replied:
Yep nice catch

peppsac commented on src/Provider/VectorTileHelper.js:
```diff
@@ -0,0 +1,77 @@
+     * @param {string} url - The URL of the tile to fetch, NOT the template: use a
+     * Provider instead if needed.
+     * @param {TileMesh} tile
+     * @param {Layer} layer
```
Maybe the documentation could be made more explicit here.
Do you really need a layer? or do you need an object with a couple of properties (.style, .filter, ... any others?)

peppsac commented on examples/planar_vector_tiles.html:
```diff
@@ -0,0 +1,94 @@
+
+                view.addLayer({
+                    type: 'color',
+                    protocol: 'xyz',
```
Maybe we could introduce a `vector-tiles` protocol and guess `protocol`, `type`, `projection` and `format` automatically since we only support 1 combination of these attributes.

peppsac replied:
I know that vector-tiles isn't a protocol, but 3dtiles isn't a protocol either. Since we currently mix protocol/format, we could take advantage of this to simplify the API.

zarov replied:
I dunno, I think it is too much a hassle, since this (declaration of layer type, protocol etc) should be more generic in the future ? And in our example we only have the `xyz` protocol, but here at IGN we have internal example with the `tms` protocol.

peppsac replied:
Agreed, but you can still assume reasonnable default values, and allow overriding by the user (in the same way as #811).

peppsac replied:
You didn't reply here?

zarov replied:
Sorry ! Yes we could but that means the introduction of a new provider if I'm not mistaken. And that doesn't make a lot of sense to me. We seem to both agree that the current `3dtiles` semantic is not the most appropriate, and that it should change, and I see this proposition for `vector-tiles` as doing a similar error like what we did for `3dtiles`.

IMHO we already have a lot of ongoing and oncoming PR refactoring the layers, providers, sources, parsers and other to take care of this later, WDYT ?

peppsac replied:
Right, let's forget about the `vector-tiles` protocol for now.

zarov replied:
Thanks !

peppsac commented on src/Parser/VectorTileParser.js:
```diff
@@ -0,0 +1,80 @@
+import { VectorTile } from '@mapbox/vector-tile';
+import GeoJsonParser from './GeoJsonParser';
+
+function readPBF(file, options) {
```
It would be nice to add at least one test for this function.

Trying to load a pbf file and verify the output features types / count would be a good start.

zarov replied:
Yep

peppsac reviewed (commented):
Thanks for updating this PR!

zarov commented:
What I find incredible here is the offset we got on the first level, see below. Is it because of the source or an error here ? 

![image](https://user-images.githubusercontent.com/741255/42682706-a07d9000-868b-11e8-887c-d5c89fe2f5ad.png)

peppsac commented on src/Provider/TMSProvider.js:
```diff
@@ -23,27 +24,34 @@ function preprocessDataLayer(layer) {
 function executeCommand(command) {
     const layer = command.layer;
     const tile = command.requester;
+    const isGlobeView = !command.view.wgs84TileLayer;
```
Why do you need this?

zarov replied:
Forgot to remove it, sorry, it was my first fix for bad tiling when using a globe view.

peppsac commented on test/unit/vectortiles.js:
```diff
@@ -0,0 +1,34 @@
+}
+
+describe('Vector tiles', function() {
+    it('should return two squares', () =>
```
:+1:

peppsac commented on examples/planar_vector_tiles.html:
```diff
@@ -0,0 +1,95 @@
+                    id: 'MVT',
+                    // eslint-disable-next-line no-template-curly-in-string
+                    url: 'https://osm.oslandia.io/data/v3/${z}/${x}/${y}.pbf',
+                    extent: [extent.west(), extent.east(), extent.south(), extent.north()],
```
You could simplify the example, and just use:
```js
url: 'https://osm.oslandia.io/data/v3/${z}/${x}/${y}.pbf',
extent,
format: 'application/x-protobuf;type=mapbox-vector',
```

(see TMSProvider.preprocessDataLayer)
```

zarov replied:
Done

peppsac commented on src/Provider/TMSProvider.js:
```diff
@@ -23,27 +24,33 @@ function preprocessDataLayer(layer) {
-            result.texture = texture;
-            result.texture.coords = coordTMSParent || coordTMS;
-            result.pitch = coordTMSParent ?
+            const pitch = coordTMSParent ?
```
I think all these changes can be removed, since it's equivalent to the original code.

zarov replied:
Yep

peppsac commented on src/Provider/VectorTileHelper.js:
```diff
@@ -0,0 +1,95 @@
+            const backgroundColor = (layer.backgroundLayer && layer.backgroundLayer.paint) ?
+                new THREE.Color(layer.backgroundLayer.paint['background-color']) :
+                undefined;
+            if (features) {
```
Suggestion: remove the `if (features)` from here and move it to `Feature2Texture.createTextureFromFeature` to avoid duplicating the backgroundColor handling?

WDYT?

zarov replied:
Yep, this could even be moved in another PR adding the `backgroundColor` functionnality ? Or I will at least create a new commit for it.

zarov replied:
Done in 4f2987c

zarov committed:
feat(core): add vector tile loading in TMS

Vector tiles can be loaded in TMS as textures. Only Mapbox Vector Tile
is supported for now. This commit adds a new Parser (VectorTileParser),
and a new Helper (similar to OGCWebServiceHelper) VectorTileHelper.

Modifications have been done in Feature2Texture to support style as a
function, and as an object.

Also add two examples showing the use of Mapbox Vector Tiles in iTowns.
The first one is loading a full vector tile layer in a Planar view,
while the second only display national parks on top of a simple Ortho
layer in a a Globe view.

zarov committed:
feat(renderer): add default background color to texture

Add a backgroundColor argument to Feature2Texture, allowing to create a
simple mono color texture if the collection argument of the function is
undefined. Otherwise, it creates an empty transparent Texture.

zarov commented:
@peppsac rebased to the last test changes.

peppsac commented on src/Renderer/ThreeExtended/Feature2Texture.js:
```diff
@@ -28,7 +39,6 @@ function drawPolygon(ctx, vertices, indices, origin, scale, properties, style =
             _lineTo(ctx, vertices[indice.offset + j], scale, origin);
         }
     }
-    ctx.closePath();
```
Are you sure about this removal?

Maybe it's not needed because the polygons we draw are guaranteed to be closed? (but I'd keep it to be on the safe side anyway)...

zarov replied:
Yes this is a fix needed because not closing the path would create visual artifacts for `multipolygon`. I missed that when working on the geojson refactoring last time, sorry about that.

peppsac reviewed (approved):
There's still a lot of work to be done to fully support vector-tiles but this PR is a great start!
